util/flexpipeviewer.py

```diff
@@ -0,0 +1,205 @@
+    else:
+        coloring_fnct = no_colorize_pipeline
+
+    with open(input_file, "r") as f:
+        for line in f:
+            line = line.strip().split(";")
+            if line[0] == "pipe":
+                # decode the instruction info
+                seq_num = int(line[1])
+                pc = int(line[2], 16)
+                upc = int(line[3], 16)
+                inst = line[4].strip()
+                creationTick = int(line[5])
+                decodeTick = int(line[6])
+                issueTick = int(line[7])
+                beginExecuteTick = int(line[8])
+                effAddredTick = int(line[9])
+                beginMemoryTick = int(line[10])
+                completionTick = int(line[11])
+                commitTick = int(line[12])
+                squashTick = int(line[13])
+
+
+                # turn the info into viewable pipeline
+                if squashTick == NEG_ONE:
+                    pipe = ["."] * (width)
+                else:
+                    pipe = ["="] * (width)
+                curr_tick = creationTick - creationTick % width_ticks
+                pipe[(creationTick - curr_tick) // cycle_time] = "c"
+                for tick, name in zip([creationTick, decodeTick, issueTick,
+beginExecuteTick, effAddredTick, beginMemoryTick, completionTick, commitTick,
+squashTick], stages):
+                    if not tick == NEG_ONE:
+                        pipe[((tick - curr_tick) // cycle_time) % width] = name
+                start_pos = ((creationTick - curr_tick) // cycle_time) % width
+                end_tick = max(creationTick, decodeTick, issueTick,
+beginExecuteTick, effAddredTick, beginMemoryTick, completionTick, commitTick,
+squashTick)
+                end_pos = ((end_tick - curr_tick) // cycle_time + 1) % width
+                pipe = coloring_fnct(pipe, start_pos, end_pos,
+                                     bg_color_char_lambda)
+                pipe.insert(0, "[")
+                pipe.append("]")
+                if seq_num in dep:
+                    dependencies = ",".join(list(map(str, dep[seq_num])))
+                else:
+                    dependencies = ""
+
+                #TODO: pc of last CALL_NEAR_I microop is incorrect?
+                info = "%6d %s (%10d) 0x%08x:%d %s %s\n" % (seq_num,
+"".join(pipe), curr_tick, pc, upc, dependencies.ljust(12), inst)
+                if seq_num in insts:
+                    insts[seq_num].append(info)
+                else:
+                    insts[seq_num] = [info]
+            elif line[0] == "dep":
+                seq_num = int(line[1])
+                if not line[1] in dep:
+                    dep[int(line[1])] = [int(line[2])]
+                else:
+                    dep[int(line[1])].append(int(line[2]))
+
+    max_seq_num = max(list(insts.keys()))
+
+    with open(output_file, "w") as g:
+        g.write(", ".join(["%s: %s" % (name, full_name)
+for full_name, name in zip(["creation", "decode", "issue",
```

Comment:
There must be a better way... :(

Reply:
> Great so far!
> 
> IIRC, you asked about what to do if two stages happen at the same time. My suggestion would be to use a new glyph to represent that. We could use some weird unicode things (e.g., æ). However, this might not be compatible with all terminals. The other option would be to choose some other letters to represent multiple stages in one cycle.

On second thought, I think it's only worth it if there are stages that frequently occur in one tick; having too many legends would make it harder to read the output.